### DISCUSSION TITLE
ICX Maker Bot (BTC)

### DISCUSSION THREAD

**Discussion by izzycsy** [DeFiCh/app#998] [Experimental]
For users who would like to test ICX (InterChain Exchange) maker bot independently on their local repository, here’s a guide. For the purpose of this test exercise, Taker (the user), will be interacting with Maker (bot) to perform the transaction.

Before starting the test, ensure that you have:
- node synced with **Testnet** and enabled `spv=1` in defi.conf file
- sufficient utxoDFI, tokenDFI and dBTC token balance

After testing, since you’re using BTC Maker bot, you should receive BTC token for dBTC. Whereas if you’re using BTC Maker bot, you should receive BTC for dBTC token. **Essentially, BTC Maker is selling BTC.**

## Prerequisite Parameters
Tips:
- Use Code Editor tool to take note the following parameters as you’ll be reusing several scripts
- Take note of the punctuation symbols, it should be printed as `' '` or `" "`, **not** `“ ”`

#### Open `defi.conf` in a Code Editor and retrieve the following parameters:
Mac — Go to Folder:
`/Users/<username>/.defi/defi.conf`

Windows:
`C:\Users\<username>\.defi\defi.conf`

<img width="892" alt="Hidden defi folder" src="https://user-images.githubusercontent.com/65014479/126188538-2eb56cf7-4302-46dd-ad31-022733e5f96c.png">

#### Check in `defi.conf` that spv is enabled: `spv=1`
```
rpcaddress (rpcbind)=127.0.0.1
rpcport=8555
rpcuser=FXdChSOx
rpcpassword=1e4fb8343d563d6ce26a035dece3b4e03db407e433e8de8f9fd7e55c820edc399
```

#### From your DeFi Wallet, retrieve the following parameters:
```
dfi_address=tqEw2Q9yGJ1UWCXdQcUwkwSX48mkvdV2sS
spv_btc_address=tb1q2pysdzz8667dv4qvptqnzvm07vw9purg0hwhy5
spv_getaddresspubkey=03a4728fae10976ad7b2288782e083d2204b16c0b631f3c82f37e22cbeae18zy70
```

- If you do not have an spvWallet Bitcoin address, run the following command in DeFi Wallet CLI
`spv_getnewaddress` in DeFi CLI

<img width="216" alt="Deno #3 spv_getnewaddress" src="https://user-images.githubusercontent.com/65014479/125783881-d1e47778-cd8e-4c78-a4b7-fd411399b41c.png">

- To retrieve your spvWallet public key, run the following command in DeFi Wallet CLI
`spv_getaddresspubkey <space> spvAddress` in DeFi CLI

<img width="583" alt="Deno #4 spv_getaddresspubkey" src="https://user-images.githubusercontent.com/65014479/125783902-79a5e181-e735-49c4-b84a-4e6247d82215.png">

## Run DeFi Wallet in the background
- Ensure that DeFi Wallet is 100% synced, and keep it running in the background

<img width="1009" alt="Deno #4 Run DeFi Wallet" src="https://user-images.githubusercontent.com/65014479/125608443-2dc2d408-f4d8-4f79-abf4-9ecb5e74dc8c.png">

## Prepare the script
- Check connection is running smoothly and prerequisites are correct
- Run your Terminal, change directory to `bin/defi-cli` or `bin/defi-cli.exe`
- You may find the latest release here: <https://github.com/DeFiCh/ain/releases>

#### Check that it’s running smoothly by using the prerequisites parameters followed by the command `getblockchaininfo`

Example script:
```
./defi-cli -testnet -rpcport=8555 -rpcuser="FXdChSOx" -rpcpassword="1e4fb8343d563d6ce26a035dece3b4e03db407e433e8de8f9fd7e55c820edc399" getblockchaininfo
```

Example output:
```
% cd defichain-1.7.11
defichain-1.7.11 % cd bin
bin % ./defi-cli -testnet -rpcport=8555 -rpcuser="FXdChSOx" -rpcpassword="1e4fb8343d563d6ce26a035dece3b4e03db407e433e8de8f9fd7e55c820edc399" getblockchaininfo
{
  "chain": "test",
  "blocks": 444247,
  "headers": 444247,
  "bestblockhash": "5c8f1218399w1c56a10f6dc8a6a146821e4ab0646dd3b451931f5e2032581a8f",
  "difficulty": 31586510.68861501,
  "mediantime": 1625830658,
  "verificationprogress": 0.9999937812653793,
  "initialblockdownload": false,
  "chainwork": "0000000000000000000000000000000000000000000002ae8c684c14dcc3ab46",
  "size_on_disk": 229772755,
  "pruned": false,
  .
  .
  .
  "eunos": {
      "type": "buried",
      "active": true,
      "height": 354950
    }
  },
  "warnings": ""
}
```

## To run BTC maker bot
**User role:** Taker
**Type:** EXTERNAL
> `EXTERNAL` would mean that the order is from BTC to DeFiChain network, while `INTERNAL` order is from DeFiChain to BTC network. 
- At the start of each command, you’ll have to use the following script followed by the respective command(s)

Example script:
```
./defi-cli -testnet -rpcport=8555 -rpcuser="FXdChSOx" -rpcpassword="1e4fb8343d563d6ce26a035dece3b4e03db407e433e8de8f9fd7e55c820edc399" <command here>
```

#### 1. Get orders
```
./defi-cli -testnet -rpcport=8555 -rpcuser="FXdChSOx" -rpcpassword="..." icx_listorders
```

Example	output
```
{
  "WARNING": "ICX and Atomic Swap are experimental features. You might end up losing your funds. USE IT AT YOUR OWN RISK.",
  "0e7762468c6213e30ac17a5ce20116621f94dc1fdb9c7a427959eb5e2c611640": {
    "status": "OPEN",
    "type": "EXTERNAL",
    "chainFrom": "BTC",
    "tokenTo": "BTC",
    "ownerAddress": "7Jw72Q9yGJ1UWCXdQcUwkwSX48mkvdV2sS",
    "amountFrom": 0.00174100,
    "amountToFill": 0.00174100,
    "orderPrice": 1.00000000,
    "amountToFillInToAsset": 0.00174100,
    "height": 468723,
    "expireHeight": 473723
  }
}
```

#### 2. Taker makes an offer from `type: EXTERNAL` output
Parameters:
orderTx: `0e77...` from **output 1’s** `"type": EXTERNAL` output
ownerAddress: <taker’s DFI address>
amount: less than "amountFROM"
expiry: `240` refers to 240 minutes; 2 hours
```
% ./defi-cli -testnet -rpcport=8555 -rpcuser="FXdChSOx" -rpcpassword="..." icx_makeoffer '{"orderTx":"0e7762468c6213e30ac17a5ce20116621f94dc1fdb9c7a427959eb5e2c611640","amount":0.0001,"ownerAddress":"tqEw2Q9yGJ1UWCXdQcUwkwSX48mkvdV2sS","receivePubkey":"03a4728fae10976ad7b2288782e083d2204b16c0b631f3c82f37e22cbeae18zy70","expiry":240}'
```

Example output

```
{
  "WARNING": "ICX and Atomic Swap are experimental features. You might end up losing your funds. USE IT AT YOUR OWN RISK.",
  "txid": "95873cdf935157b9c0395d307c9b40a9bab30a3057a548b3df2678373479437a"
}
```

#### 3. After a few minutes, check if maker accepted taker’s offer
Parameters:
txid: from **output 2**
```
% ./defi-cli ... icx_listhtlcs '{"offerTx":"95873cdf935157b9c0395d307c9b40a9bab30a3057a548b3df2678373479437a"}'
```

Example output
```
{
  "WARNING": "ICX and Atomic Swap are experimental features. You might end up losing your funds. USE IT AT YOUR OWN RISK.",
  "53a7a2fc73c004af3cc51199e2041d6c44c05c621ef81174e85ce61e6ee9c023": {
    "type": "EXTERNAL",
    "status": "OPEN",
    "offerTx": "95873cdf935157b9c0395d307c9b40a9bab30a3057a548b3df2678373479437a",
    "amount": 0.00010000,
    "amountInDFCAsset": 0.00010000,
    "hash": "ded4a16fde0a2dbab8db98c55d29733f40506068aaad84387a201ce53657465d",
    "htlcScriptAddress": "2NDa25zJ9oe6nMAdRUkuoaLARKjm2b37Dv1",
    "ownerPubkey": "03668b5dc4f33dab92cd5b70c034a88e0d2510b14e39f3995245acb4d8723c2b35",
    "timeout": 80,
    "height": 471592
  }
}
```

#### 4. With the above output, create HTLC (Hashed Timelock Contract) on SPV
Parameters (separate with a space):
Param 1 (the one who receives the BTC): Taker’s `receivePubKey` (spv_getaddresspubkey) from Prerequisite Parameters
Param 2: Maker’s `receivePubKey` from **output 3**
Param 3: from **output 3**, timeout should be less than the order timeout number and larger than 24
Param 4: **output 3’s** `hash`
```
% ./defi-cli ... spv_createhtlc 03a4728fae10976ad7b2288782e083d2204b16c0b631f3c82f37e22cbeae18zy70 03668b5dc4f33dab92cd5b70c034a88e0d2510b14e39f3995245acb4d8723c2b35 80 ded4a16fde0a2dbab8db98c55d29733f40506068aaad84387a201ce53657465d
```

Example output
```
{
  "address": "2NDa25zJ9oe6nMAdRUkuoaLARKjm2b37Dv1",
  "redeemScript": "63a820ded4a16fde0a2dbab8db98c55d29733f40506068aaad84387a201ce53657465d882103a4728fae10976ad7b2288782e083d2204b16c0b631f3c82f37e22cbeae18ad60670150b2752103668b5dc4f33dab92cd5b70c034a88e0d2510b14e39f3995245acb4d8723c2b3568ac"
}
```

#### 5. Check if the balance on BTC HTLC address is the same in Taker’s offer
Parameters (separate with a space):
Param 1: `1` refers to checking one result
Param 2: `htlcScriptAddress` from **output 4**
```
% ./defi-cli ... spv_listreceivedbyaddress 1 2NDa25zJ9oe6nMAdRUkuoaLARKjm2b37Dv1
```

Example output
```
[
  {
    "address": "2NDa25zJ9oe6nMAdRUkuoaLARKjm2b37Dv1",
    "type": "HTLC",
    "amount": 0.00010000,
    "confirmations": 1,
    "txids": [
      "433edf8ec7d4a1ea2e32a3d783dcb0d7cda2737922a0410dcd6757fcd20cb269"
    ]
  }
]
```

#### 6. Submit HTLC to DeFiChain
Parameters:
offerTx: from **output 3**
amount: from **output 2**
hash: from **output 3**
```
% ./defi-cli ... icx_submitdfchtlc '{"offerTx":"95873cdf935157b9c0395d307c9b40a9bab30a3057a548b3df2678373479437a","amount":0.0001,"hash":"ded4a16fde0a2dbab8db98c55d29733f40506068aaad84387a201ce53657465d"}'

```

Example output
```
{
  "WARNING": "ICX and Atomic Swap are experimental features. You might end up losing your funds. USE IT AT YOUR OWN RISK.",
  "txid": "fd0ff063343ff82c73f26690188949e6f3539fbabdfe475aac34e73ee0266251"
}
```

#### 7. Taker checks constantly if Maker claims the DeFiChain HTLC
Parameters:
offerTx: from **output 3**
```
% ./defi-cli ... icx_listhtlcs '{"offerTx":"95873cdf935157b9c0395d307c9b40a9bab30a3057a548b3df2678373479437a"}'
```

Example output
```
{
  "WARNING": "ICX and Atomic Swap are experimental features. You might end up losing your funds. USE IT AT YOUR OWN RISK.",
  "fbd3b5a1eea6738337a64e54e044f31afc943293defd32a5dd0409b532a9eeef": {
    "type": "CLAIM DFC",
    "dfchtlcTx": "fd0ff063343ff82c73f26690188949e6f3539fbabdfe475aac34e73ee0266251",
    "seed": "604a6b50d453ef9af5a58c4bd0adbb290c71cbbd038e6a038e5e911a84b7961c",
    "height": 471606
  }
}
```

#### 8. Taker to use seed to claim SPV HTLC (BTC)
Parameters (separate with <space>):
Param 1: `htlcScriptAddress` from **output 4**
Param 2: Taker’s spvAddress
Param 3: `seed` from **output 7**
```
% ./defi-cli ... spv_claimhtlc "2NDa25zJ9oe6nMAdRUkuoaLARKjm2b37Dv1" "tb1q2pysdzz8667dv4qvptqnzvm07vw9purg0hwhy5" "604a6b50d453ef9af5a58c4bd0adbb290c71cbbd038e6a038e5e911a84b7961c"
```

Example output
```
{
  "txid": "7fdfe6d42053dffe6ceec05eede4a1ad1a18bb1919cedcdfe4077fdc3f32df71",
  "sendmessage": "Undefined error: 0"
}
```

You may check your transaction on Bitcoin Testnet Explorer: <https://www.blockchain.com/explorer?view=btc-testnet>
<img width="708" alt="ICX #8 BTC check txid" src="https://user-images.githubusercontent.com/65014479/126456442-034b2d1e-0380-4b0b-9e3a-f30d46d9ea76.png">

#### Footnote
- To retrieve latest version for defi-cli: go to <https://github.com/DeFiCh/ain/releases>
- To retrieve latest version for DeFi Wallet: go to <https://github.com/DeFiCh/app/releases>
- Warning: ICX and Atomic Swap are experimental features. You might end up losing your funds. USE IT AT YOUR OWN RISK.

**Comment by salomonkarsten** [comment]
On point 3 can you explain how maker accepts an offer. What command to use? It seems nothing happening with command.

icx_listhtlcs {"offerTx":"txid"}
{
  "WARNING": "ICX and Atomic Swap are experimental features. You might end up losing your funds. USE IT AT YOUR OWN RISK."
}
